padl/utils.py

```diff
@@ -61,3 +63,122 @@ def __getattr__(self, attr):
 
 #: Transform factory for capturing attributes/ get-items.
 same = _Same()
+
+
+class _Debug:
+    """Customized debugger for :class:`padl.Transform`s. When an exception on the execution of
+    a :class`Transform` is produced and a :class:`_Debug` object is called, an interactive debugger
+    at different levels in the :class:`Transform` is gotten.
+
+    At the top, the user interacts with the entire transform and its absolute input. One level
+    down, it goes directly to the stage that got the Exception (either to
+    :meth:`padl.transforms.Transform.pd_preprocess`, :meth:`padl.transforms.Transform.pd_forward`,
+    or :meth:`padl.transforms.Transform.pd_postprocess`) and each level deeper moves recursively
+    inside the element that failed until the :class:`AtomicTransform` that got the Exception is
+    reached.
+    """
+    def __init__(self):
+        self.trans = None
+        self.args = None
+        self.default_msg = (
+            'Defined commands are: \n'
+            '   u(p): step up\n'
+            '   d(own): step down\n'
+            '   w(here am I?): show code position\n'
+            '   i(nput): show input here\n'
+            '   r(epeat): repeat here (will produce the same exception)\n'
```

Comment:
I think we should consider removing the `r(epeat)` option since it currently requires us to store the original input arguments, which isn't ideal. Also looking into the `pdb` [debugger](https://www.uni-muenster.de/AMM/num/Vorlesungen/Pythonkurs_SS15/pdb_cheatsheet.pdf), I think this option is non-standard. What do you think @wuhu @eramdiaz ?

Reply:
I see it as a nice option to check again the error that occurred after inspecting inside the transform and making an hypothesis of the error. By the original input arguments, you refer to the absolute inputs to the transform, right? Keeping the arguments is also needed for the `i(nput)` option, which is very clarifying in my opinion.

Reply:
> I see it as a nice option to check again the error that occurred after inspecting inside the transform and making an hypothesis of the error. 

I'm not sure I agree with this since I think you usually see the original error before entering a debugger right? 

> By the original input arguments, you refer to the absolute inputs to the transform, right? 

Yes

> Keeping the arguments is also needed for the `i(nput)` option, which is very clarifying in my opinion.

Yes I could see how this could be clarifying. However, isn't it possible to see what the inputs are through the traceback? For example, progressing upwards through the call tree up until the first call and printing the input variable?

Reply:
mmm, honestly I don't know about that, maybe @wuhu has an idea about this. This may affect also to the in-line compiler at the debugger: at the moment we can give an standard input to the debugger like:
```
> i.shape
torch.Size([2, 3, 8]) # for example
```
I am not sure if this would work without storing the args

Reply:
Okay sounds good and I agree we are just delaying garbage collection.